spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkPositionDeltaWrite.java

```diff
@@ -0,0 +1,461 @@
+    }
+
+    private void commitOperation(SnapshotUpdate<?> operation, String description) {
+      LOG.info("Committing {} to table {}", description, table);
+      if (applicationId != null) {
+        operation.set("spark.app.id", applicationId);
+      }
+
+      extraSnapshotMetadata.forEach(operation::set);
+
+      if (wapEnabled && wapId != null) {
+        // write-audit-publish is enabled for this table and job
+        // stage the changes without changing the current snapshot
+        operation.set(SnapshotSummary.STAGED_WAP_ID_PROP, wapId);
+        operation.stageOnly();
+      }
+
+      long start = System.currentTimeMillis();
+      operation.commit(); // abort is automatically called if this fails
+      long duration = System.currentTimeMillis() - start;
+      LOG.info("Committed in {} ms", duration);
+    }
+  }
+
+  public static class DeltaTaskCommit implements WriterCommitMessage {
+    private final DataFile[] dataFiles;
+    private final DeleteFile[] deleteFiles;
+    private final CharSequence[] referencedDataFiles;
+
+    DeltaTaskCommit(DeleteWriteResult result) {
+      this.dataFiles = new DataFile[0];
+      this.deleteFiles = result.deleteFiles().toArray(new DeleteFile[0]);
+      this.referencedDataFiles = result.referencedDataFiles().toArray(new CharSequence[0]);
+    }
+
+    DataFile[] dataFiles() {
+      return dataFiles;
+    }
+
+    DeleteFile[] deleteFiles() {
+      return deleteFiles;
+    }
+
+    public CharSequence[] referencedDataFiles() {
+      return referencedDataFiles;
+    }
+  }
+
+  private static class PositionDeltaWriteFactory implements DeltaWriterFactory {
+    private final Broadcast<Table> tableBroadcast;
+    private final Context ctx;
+
+    PositionDeltaWriteFactory(Broadcast<Table> tableBroadcast, Context ctx) {
+      this.tableBroadcast = tableBroadcast;
+      this.ctx = ctx;
+    }
+
+    @Override
+    public DeltaWriter<InternalRow> createWriter(int partitionId, long taskId) {
+      Table table = tableBroadcast.value();
+      Map<Integer, PartitionSpec> specs = table.specs();
+      FileIO io = table.io();
+
+      OutputFileFactory deleteFileFactory = OutputFileFactory.builderFor(table, partitionId, taskId)
+          .format(ctx.deleteFileFormat())
+          .build();
+
+      SparkFileWriterFactory writerFactory = SparkFileWriterFactory.builderFor(table)
+          .dataFileFormat(ctx.dataFileFormat())
+          .dataSchema(ctx.dataSchema())
+          .dataSparkType(ctx.dataSparkType())
+          .deleteFileFormat(ctx.deleteFileFormat())
+          .positionDeleteSparkType(ctx.deleteSparkType())
+          .build();
+
+      Types.StructType partitionType = Partitioning.partitionType(table);
+      StructType sparkPartitionType = (StructType) SparkSchemaUtil.convert(partitionType);
+      InternalRowWrapper partitionRowWrapper = new InternalRowWrapper(sparkPartitionType);
+
+      // build projections that would project needed partitions fields for each spec
```

Comment:
It would be nice to have a bit more detail about what's happening here: the projection is from the union partition type from all specs to a specific partition types for the given spec.

Reply:
Added a comment.